### ISSUE TITLE
Depending on slf4j causes test worker throw ClassCastException

### ISSUE THREAD

**Issue by blindpirate** [gradle/gradle#4981]
This issue has been existing at least since Gradle 2.0.

In our test worker, the [classloader hierarchy](https://github.com/gradle/gradle/blob/0437bb59d6d816797e64e7f13ade96bad198bae3/subprojects/core/src/main/java/org/gradle/process/internal/worker/child/ApplicationClassesInSystemClassLoaderWorkerImplementationFactory.java#L70) is:

```
 *                       jvm bootstrap
 *                             |
 *                             |
 *                        jvm system
 *           (GradleWorkerMain, application classes)
 *                             |
 *                             |
 *                          filter
 *                    (shared packages)
 *                             |
 *                             |
 *                       implementation
 *          (SystemApplicationClassLoaderWorker, logging)
```

If users depends on `org.slf4j:slf4j-api` (no matter which version), the slf4j jar will occurs in application classloader (i.e.  jvm system). In this case, during worker startup, [this line](https://github.com/gradle/gradle/blob/5c47e0dbfddcc57fbde89a9c6c56c9520363cd3f/subprojects/logging/src/main/java/org/gradle/internal/logging/slf4j/Slf4jLoggingConfigurer.java#L46) will throw `ClassCastException`:

```
OutputEventListenerBackedLoggerContext context = (OutputEventListenerBackedLoggerContext) LoggerFactory.getILoggerFactory();
```

`LoggerFactory` will try to load `org.slf4j.impl.StaticLoggerBinder`, in this case, `LoggerFactory` is loaded by application class loader, but Gradle's `org.slf4j.impl.StaticLoggerBinder` implementation exists in "implementation" layer in the graph above, therefore, `LoggerFactory.getILoggerFactory()` returns a fallback `org.slf4j.helpers.NOPLoggerFactory` and the `ClassCastException` is thrown.

It seems a bit hard to keep all Gradle logging infrastructure working in this case. However, at least we shouldn't allow the exception thrown and worker process crash.

### Steps to reproduce

See samples in these bug report: #4974  #2657

**Comment by kristian-elder** [comment]
This is currently a blocker for us upgrading to Java 9+ / Jigsaw. Is any workaround possible at all, short of not using slf4j at all?

**Comment by oehme** [comment]
Duplicate of #2657
